src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -786,3 +788,54 @@ case class TableAggregateByKey(child: TableIR, expr: IR) extends TableIR {
     prev.copy(rvd = newRVD, typ = typ)
   }
 }
+
+case class TableOrderBy(child: TableIR, sortFields: IndexedSeq[SortField]) extends TableIR {
+  // TableOrderBy expects an unkeyed child, so that we can better optimize by
+  // pushing these two steps around as needed
+  require(child.typ.key.isEmpty)
```

Comment:
I'm confused by this.  Why not just have a TableOrderBy by unkeyed?

Also, this is likely failing, since Python doesn't unkey before calling Table.orderBy.

Reply:
As I've thought about the optimization passes, it seems like it makes sense to keep separate things separate. for instance, push an unkey down to a read so we never have to decode + depend on keys. If we have Table order by do both, then that's harder

Reply:
Is that wrong?

Reply:
This also means we eliminate order by nodes that aren't needed without inserting unkey nodes

Reply:
No, that's true.  There's also a tension with composability of the IR: you've now constrained the the child of this node even in cases when it is "sensible" and that is likely to lead to bugs.  (I guess that's why this PR is failing now.)  In my mind, ordering and keying are inextricably linked, and there a small set of IR nodes that deal with ordering: KeyBy and OrdrerBy (I generally think Unkey should just be KeyBy(None)).  Then optimizations eliminating unnecessary orderings will have to deal with these (two) nodes.

If you still think it's better this way, leave it.